### PR TITLE
Merge create and initialize to limit data races

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Resolves: #2341

## Description

Avoid inserting partially initialized objects into the global object tracking table.

Previously, link and program objects where first inserted into the global object tracking table, then initialized which results in a race between threads reading the objects via calls to reference object by id and the code performing the initialization. This change alters the order so that objects are first initialized and then inserted into the global table to limit the possible data races that can occur.

## Testing

CI/CD

## Documentation

No.

### CODE REVIEW

File: libs/execution_context/ebpf_program.c [hunk 1]
```diff
@@ -447,79 +451,38 @@ ebpf_program_create(_Outptr_ ebpf_program_t** program)
...
             "Program name must be less than BPF_OBJ_NAME_LEN",
             program_parameters->program_name.length);
-        return_value = EBPF_INVALID_ARGUMENT;
+        retval = EBPF_INVALID_ARGUMENT;
         goto Done;
     }
 
-    return_value = ebpf_duplicate_utf8_string(&local_program_name, &program_parameters->program_name);
-    if (return_value != EBPF_SUCCESS) {
+    retval = ebpf_duplicate_utf8_string(&local_program_name, &program_parameters->program_name);
```
Comment (shankarseal) [inline libs/execution_context/ebpf_program.c:466]: Do you need to change the SAL annotation that program_parameters need not be protected by the lock?

Comment (Alan-Jowett) [reply libs/execution_context/ebpf_program.c:466]: That's why I filed #2325. The exact rules about when it's safe to access fields in ebpf_program_t is unclear. Some fields are protected by ebpf_program_t::lock and others are protected by the rundown protection on the program information provider. Until we document and audit the carefully, we can't be sure it's race free.

**Review by dthaler — DISMISSED** [review]

**Review by dv-msft — DISMISSED** [review]

File: libs/execution_context/ebpf_program.c (hunk 1 above)
Comment (shankarseal) [reply libs/execution_context/ebpf_program.c:466]: But won't SAL complain about the changes you made in this PR?

File: libs/execution_context/ebpf_program.c [hunk 2]
```diff
@@ -528,38 +491,45 @@ ebpf_program_initialize(_Inout_ ebpf_program_t* program, _In_ const ebpf_program
...
-    ebpf_lock_unlock(&program->lock, state);
-    lock_held = false;
+    retval = _ebpf_program_load_providers(local_program);
+    if (retval != EBPF_SUCCESS) {
+        goto Done;
+    }
 
-    return_value = _ebpf_program_load_providers(program);
-    if (return_value != EBPF_SUCCESS) {
+    retval = ebpf_object_initialize(
```
Comment (shankarseal) [inline libs/execution_context/ebpf_program.c:512]: I would also put a comment in `ebpf_object_initialize` that says any objects that are "derived" structs of `ebpf_object` must call this function only after the structure is completely initialized.

**Review by shankarseal — APPROVED** [review]

File: libs/execution_context/ebpf_program.c (hunk 2 above)
Comment (saxena-anurag) [inline libs/execution_context/ebpf_program.c:512]: same here: adding a comment that this needs to be the last step.

File: libs/execution_context/ebpf_program.c (hunk 1 above)
Comment (Alan-Jowett) [reply libs/execution_context/ebpf_program.c:471]: Doesn't appear to.

File: libs/execution_context/ebpf_program.c (hunk 2 above)
Comment (Alan-Jowett) [reply libs/execution_context/ebpf_program.c:533]: Fixed,

File: libs/execution_context/ebpf_link.c [hunk 3]
```diff
@@ -89,58 +89,57 @@ _ebpf_link_free(_Frees_ptr_ ebpf_core_object_t* object)
...
-        memcpy(link->client_data.data, context_data, context_data_length);
+        memcpy(local_link->client_data.data, context_data, context_data_length);
     }
 
-    link->attach_type = attach_type;
+    local_link->attach_type = attach_type;
+
+    ebpf_lock_create(&local_link->attach_lock);
+
+    ebpf_result_t result = ebpf_object_initialize(&local_link->object, EBPF_OBJECT_LINK, _ebpf_link_free, NULL);
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply libs/execution_context/ebpf_link.c:129]: Fixed.

File: libs/execution_context/ebpf_program.c (hunk 2 above)
Comment (Alan-Jowett) [reply libs/execution_context/ebpf_program.c:533]: Fixed

**Review by saxena-anurag — APPROVED** [review]

**Pull request merged**